diff --git a/document.js b/document.js
index 7c0f50dc53a02bd432788099748f6a87b5d64b3c..49696b86ca51a852a8e9271c24075447a1a675d1 100644
--- a/document.js
+++ b/document.js
@@ -1,6 +1,5 @@
 var topLevel = typeof global !== 'undefined' ? global :
     typeof window !== 'undefined' ? window : {}
-var minDoc = require('min-document');
 
 var doccy;
 
@@ -10,7 +9,7 @@ if (typeof document !== 'undefined') {
     doccy = topLevel['__GLOBAL_DOCUMENT_CACHE@4'];
 
     if (!doccy) {
-        doccy = topLevel['__GLOBAL_DOCUMENT_CACHE@4'] = minDoc;
+        doccy = topLevel['__GLOBAL_DOCUMENT_CACHE@4'] = {};
     }
 }
 
diff --git a/package.json b/package.json
index 052726a208918fc64397e23e7e36501666c8fb5f..8909479ab2900f4664f228ad4951081246fc19e8 100644
--- a/package.json
+++ b/package.json
@@ -17,11 +17,9 @@
     "email": "raynos2@gmail.com"
   },
   "browser": {
-    "min-document": false,
     "individual": false
   },
   "dependencies": {
-    "min-document": "^2.19.0",
     "process": "^0.11.10"
   },
   "devDependencies": {
